src/conn.ts

```diff
@@ -1,183 +1,376 @@
+import { states } from 'minecraft-protocol';
+import { getLoginSequencePackets } from './packets';
+const bufferEqual = require('buffer-equal');
 
 export type Packet = [name: string, data: any];
 
 export type ClientEventTuple = [event: string, listener: (...args: any) => void];
 export type ClientEvents = (ClientEventTuple | ((conn: Conn, pclient: Client) => ClientEventTuple))[];
 
 export type Client = mcpClient & {
-  //* whitelist overwrites not being the main pclient
-  toServerWhiteList?: string[];
-  //* filter for when the client is the main client
-  toServerBlackList?: string[];
-  //* filter when the client is attached
-  toClientBlackList?: string[];
+  toClientMiddlewares: PacketMiddleware[];
+  toServerMiddlewares: PacketMiddleware[];
+
+  on(event: 'mcproxy:detach', listener: () => void): void;
+  on(event: 'mcproxy:heldItemSlotUpdate', listener: () => void): void;
 };
 
 export class ConnOptions {
-  events: ClientEvents = [];
-  //* a whitelist for the internal bot
-  internalWhitelist: string[] = ['keep_alive'];
-  //* filter for the main client as long as it is not overwritten by the client itself
-  toServerBlackList: string[] = ['keep_alive'];
-  //* filter for all attached clients if it is not overwritten by the client
-  toClientBlackList: string[] = ['keep_alive'];
+  optimizePacketWrite: boolean = true;
+  //* Middleware to control packets being send to the client and server
+  toClientMiddleware?: PacketMiddleware[] = [() => {}];
+  //* Middleware to control packets being send to the client and server
+  toServerMiddleware?: PacketMiddleware[] = [() => {}];
+}
+
+export interface PacketCanceler {
+  /** Has property .isCanceled: boolean indicating if the packet has been canceled by another middleware.
+   * Use `cancel(false)` to un-cancel the packet again.
+   */
+  (unCancel?: boolean): void;
+  isCanceled: boolean;
+}
+
+export interface packetUpdater {
+  (update?: boolean): void;
+  isUpdated: boolean;
+}
+
+/**
+ * Middleware manager for packets. Used to modify cancel or delay packets being send to the client or server.
+ */
+export interface PacketMiddleware {
+  /** Contains meta information about the packet that is triggered. See the properties for more info. */
+  (
+    info: {
+      /** Direction the packet is going. Should always be the same direction depending on what middleware direction you
+       * are registering.
+       */
+      bound: 'server' | 'client';
+      /** Only 'packet' is implemented */
+      writeType: 'packet' | 'rawPacket' | 'channel';
+      /** Packet meta. Contains the packet name under `name` also see {@link PacketMeta} */
+      meta: PacketMeta;
+    },
+    /** The client connected to this packet */ pclient: Client,
+    /** Parsed packet data as returned by nmp */ data: any,
+    /** A handle to cancel a packet from being send. Can also force a packet to be send */ cancel: PacketCanceler,
+    /** Indicate that a packet has been modified */ update: packetUpdater
```

Comment:
I think the cancel and update api is still a bit lacklustre, i would prefer it being controlled be the return type of the function. (I guess this is the chaining kind of Middleware again?)
```suggestion
    // The client connected to this packet
    pclient: Client,
    // Parsed packet data as returned by nmp
    data: any,
    // A handle to cancel a packet from being send. Can also force a packet to be send
    cancel: PacketCanceler,
    // Indicate that a packet has been modified
    update: packetUpdater
```

Reply:
Changed